File: abstr2/urls.py
from abstr2.context import ContextMap


def is_supported_site(contextmap: ContextMap):
    return contextmap.current_url.startswith("https://") or contextmap.current_url.startswith("http://")

def does_not_need_be_indexed(contextmap: ContextMap):
    url = contextmap.current_url
    urlfilter = contextmap.url_filter
    rp = contextmap.rp
    if not (is_supported_site(contextmap)):
        return False
    if not urlfilter.matches(url):
        return False
    if rp is not None and not rp.can_fetch("snail", url):
        return False

File: src/main/java/search/Urls.java
package search;

/**
 * Port of abstr2.urls.
 * Can be instantiated for testing.
 */
public final class Urls {

    public Urls() {
    }

    public boolean isSupportedSite(ContextMap contextmap) {
        String url = contextmap.currentUrl;
        return url != null && (url.startsWith("https://") || url.startsWith("http://"));
    }

    public boolean needsBeIndexed(ContextMap contextmap) {
        String url = contextmap.currentUrl;
        URLFilter urlfilter = contextmap.urlFilter;
        RobotRules rp = contextmap.robotRules;
        if (!isSupportedSite(contextmap)) {
            return false;
        }
        if (urlfilter == null || !urlfilter.matches(url)) {
            return false;
        }
        if (rp != null && !rp.isAllowed(url)) {
            return false;
        }
        return true;
    }
}
